### PR TITLE
fix(eslint-plugin): [return-await] correct autofixer in binary expression

### LINKED ISSUE #4090: [return-await] wrong autofix of `return undefined ?? promise` to `return await undefined ?? promise`
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/await-thenable": "error"
  }
}
```

```TS
async function bar() {}
async function foo() {
	try {
		return window.foo ?? bar();
	} catch {}
}
```

tsconfig: https://github.com/sindresorhus/tsconfig/blob/67d56ca87d2790081696fe3d745befa7e9ff38b8/tsconfig.json

**Expected Result**

```TS
async function bar() {}
async function foo() {
	try {
		return window.foo ?? await bar();
		                  // ^ correct position of `await`
	} catch {}
}
```

**Actual Result**

```TS
async function bar() {}
async function foo() {
	try {
		return await await await await await await await await await await window.foo ?? bar();
		    // ^ incorrect position of `await`, which is then "autofixed" repeatedly
	} catch {}
}
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.3.0` |
| `@typescript-eslint/parser`        | `5.3.0` |
| `TypeScript`                       | `4.4.4` |
| `ESLint`                           | `8.1.0` |
| `node`                             | `16.4.2` |

### PR DESCRIPTION
**Pull request by islandryu**
## PR Checklist

-   [x] Addresses an existing issue: fixes #4090
-   [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
-   [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

### CODE REVIEW

**Comment by islandryu** [comment]
I'm sorry.
I did a forcepush by mistake.
I didn't change anything.

**Review by bradzacher — CHANGES_REQUESTED** [review]
Thanks for the PR!

This is definitely a good start - but the problem is that there are a lot of cases that it won't handle properly.

The problem we're trying to solve here is that of "precedence" - meaning we need to check if the thing on the right has a lower precedence than the `await` we're inserting. If it has a lower precedence, then it is evaluated after the await. If it has a higher precedence, then it is evaluated before the await.

Luckily this problem is already solved for us, to some degree:
https://github.com/microsoft/TypeScript/blob/af689cc5d527789cbc1d18a7b8aae6fa46051ea7/src/compiler/utilities.ts#L3446-L3783

All of this code is specifically built for TS nodes. Additionally it's all internal to TS (so we can't use it anyways). It would be a good idea for us to maintain a copy of this for ourselves using our `AST_NODE_TYPES` (and with a non-const enum). We can place this function in the `utils` package so anyone can consume it easily.

---

As an example - from the above code we can see that `BinaryExpression` has a precedence of `OperatorPrecedence.Coalesce` (or `4`), whereas `AwaitExpression` has a precedence of `OperatorPrecedence.Unary` (or `16`)

This means that `AwaitExpression` has a higher precedence than `BinaryExpression` - thus it is evaluated first.
So by evaluating the precedences, we know that before we insert an `await` we must wrap the child `BinaryExpression` in parentheses so that it is evaluated correctly.

This system handles all cases correctly and saves us having to manually handle individual cases.

**Comment by islandryu** [comment]
@bradzacher 
Thanks for the review.
Am I correct in my understanding?

``` js
// CallExpression has a higher precedence than awaitExpression, so () is not necessary.
// input
return a();

// output
return await a();


// All binaryExpression has a lower precedence than awaitExpression, so () is necessary.
// input
return a() || undefined
return undefined || a()
return undefined ?? a()

// output 
return await (a() || undefined)
return await (undefined || a())
return await (undefined ?? a())


// AssertExpression has a lower precedence than awaitExpression, so () is necessary.
// input
return a() as T

// output
return await (a() as T)
```

**Comment by bradzacher** [comment]
You are correct!
By inspecting the precedence we can determine whether or not we need parentheses at all.

File: packages/eslint-plugin/src/util/getOperatorPrecedence.ts [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+    case SyntaxKind.PercentToken:
+      return OperatorPrecedence.Multiplicative;
+    case SyntaxKind.AsteriskAsteriskToken:
+      return OperatorPrecedence.Exponentiation;
+  }
+
+  // -1 is lower than all other precedences.  Returning it will cause binary expression
+  // parsing to stop.
+  return -1;
+}
```
Comment (islandryu) [inline packages/eslint-plugin/src/util/getOperatorPrecedence.ts:347]: @bradzacher 
Precedence now determines whether parentheses need to be added or not.

I copied the typescript code verbatim, is that ok?
https://github.com/microsoft/TypeScript/blob/af689cc5d527789cbc1d18a7b8aae6fa46051ea7/src/compiler/utilities.ts#L3446-L3783

File: packages/eslint-plugin/src/util/getOperatorPrecedence.ts [hunk 2]
```diff
@@ -0,0 +1,347 @@
+import { SyntaxKind } from 'typescript';
+
+export enum OperatorPrecedence {
+  // Expression:
+  //     AssignmentExpression
```
Comment (islandryu) [inline packages/eslint-plugin/src/util/getOperatorPrecedence.ts:5]: @bradzacher
Precedence now determines whether parentheses need to be added or not.

I copied the typescript code verbatim, is that ok?
https://github.com/microsoft/TypeScript/blob/af689cc5d527789cbc1d18a7b8aae6fa46051ea7/src/compiler/utilities.ts#L3446-L3783

**Comment by islandryu** [comment]
I could not test all of getOperatorPrecedence.ts with return-await alone.

File: packages/eslint-plugin/src/util/getOperatorPrecedence.ts (hunk 2 above)
Comment (bradzacher) [reply packages/eslint-plugin/src/util/getOperatorPrecedence.ts:5]: not a problem at all - we *have* to copy it because it's a `const` enum!

**Review by bradzacher — APPROVED** [review]
this is looking good to me - this is awesome!
thanks so much for improving the fixer here and adding utils!

**Pull request merged**